src/crimson/admin/admin_socket.cc

```diff
@@ -0,0 +1,646 @@
+      return seastar::futurize_apply(
+               [this, &command, &cmdmap, &format, &out, f = ftr.get()] {
+                 return exec_command(f, command, cmdmap, format, out);
+               })
+        .then_wrapped([&br](seastar::future<> res) -> seastar::future<bool> {
+          //  we now have a ready future (or a failure)
+          if (res.failed()) {
+            std::ignore = res.handle_exception([](std::exception_ptr eptr) {
+              return seastar::make_ready_future<>();
+            });
+          } else {
+            br = true;  // exec_command() successful
+            res.ignore_ready_future();
+          }
+
+          return seastar::make_ready_future<bool>(br);
+        })
+        .then([&ftr, &out](auto res) -> seastar::future<bool> {
+          ftr->close_section();
+          ftr->enable_line_break();
+          ftr->flush(out);
+          return seastar::make_ready_future<bool>(res);
+        });
+    });
+}
+
+seastar::future<AsokRegistrationRes> AdminSocket::register_server(
+  hook_server_tag server_tag, const std::vector<AsokServiceDef>& apis_served)
+{
+  return seastar::with_lock(
+    servers_tbl_rwlock,
+    [this, server_tag, &apis_served]() -> AsokRegistrationRes {
+      auto ne = servers.try_emplace(server_tag, apis_served);
+      //  was this server tag already registered?
+      if (!ne.second) {
+        return {};
+      }
+      logger().info("{}: pid:{} ^:{} (tag: {})", __func__, (int)getpid(),
+		    (uint64_t)(this), (uint64_t)(server_tag));
+      return this->shared_from_this();
+    });
+}
+
+seastar::future<> AdminSocket::unregister_server(hook_server_tag server_tag)
+{
+  logger().debug("{}: pid:{} server tag: {})", __func__, (int)getpid(),
+                 (uint64_t)(server_tag));
+
+  return seastar::with_lock(servers_tbl_rwlock,
+                            [this, server_tag] {
+                              auto srv_itr = servers.find(server_tag);
+                              if (srv_itr == servers.end()) {
+                                logger().warn(
+                                  "unregister_server(): unregistering a "
+                                  "non-existing registration (tag: {})",
+                                  (uint64_t)(server_tag));
+                              } else {
+                                servers.erase(srv_itr);
+                              }
+                              return seastar::now();
+                            })
+    .finally([server_tag] {
+      logger().debug("unregister_server(): done removing server block {}",
+                     (uint64_t)(server_tag));
+    });
+}
+
+seastar::future<> AdminSocket::unregister_server(hook_server_tag server_tag,
+                                                 AdminSocketRef&& server_ref)
+{
+  // reducing the ref-count on us (the ASOK server) by discarding server_ref:
+  return seastar::do_with(std::move(server_ref), [this, server_tag](auto& srv) {
```

Comment:
why do you want to use `do_with()` here?

Reply:
I wish to move the ownership from the caller, but maintain the server's ref count till
after unregister_server() returns. As unregister_server() does not need that ref as a
parameter, I did not see a cleaner way to perform that (and do_with() is made for this use).
Open to suggestions, if do_with() is wrong.